### PR TITLE
udp: update `tokio_udp::UdpFramed` to std::future

### LINKED ISSUE #1363: Can't use codec and Framed with Udp
## Version
Master as of today (Jul 27 2019)

## Platform
Linux, not that it should matter

## Subcrates
tokio-udp, tokio-codec

## Description
In tokio 0.1, you could use a `UdpFramed<SomeCodec>` to have a codec applied to your UDP socket. In tokio master, this seems to be replaced with `Framed`, which only supports `AsyncRead + AsyncWrite` types, which `UdpSocket` is not. 

Is there any planned way to use `Framed` with UDP? I'm willing to help if I'm shown in the right direction.

### PR DESCRIPTION
**Pull request by JohnDoneth**
- [x] Refactor methods in `UdpSocket` that used to take `self: Pin<&mut>` to take `&mut self` instead.
- [x] Update `UdpFramed` to std::future.
- [x] Update `UdpFramed` tests.
- [x] Update udp-codec example to use `UdpFramed`.

Closes #1363.

### CODE REVIEW

**Comment by JohnDoneth** [comment]
@carllerche I've fixed the merge conflicts. Looks like the echo_server test in mio is failing on FreeBSD for some reason though.
```
failures:
---- echo_server stdout ----
thread 'echo_server' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `16384`', /.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.6.19/src/sys/unix/kqueue.rs:122:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
failures:
    echo_server
```

**Comment by DoumanAsh** [comment]
@JohnDoneth it is known problem that need to be fixed on libc side

File: tokio-udp/src/frame.rs [hunk 1]
```diff
@@ -33,79 +36,98 @@ pub struct UdpFramed<C> {
...
             (n, addr)
         };
         trace!("received {} bytes, decoding", n);
-        let frame_res = self.codec.decode(&mut self.rd);
-        self.rd.clear();
+        let frame_res = pin.codec.decode(&mut pin.rd);
+        pin.rd.clear();
         let frame = frame_res?;
         let result = frame.map(|frame| (frame, addr)); // frame -> (frame, addr)
+        let result = result.unwrap();
```
Comment (LucioFranco) [inline tokio-udp/src/frame.rs]: why is there an unwrap here?

Comment (JohnDoneth) [reply tokio-udp/src/frame.rs]: Ah, my mistake, I forgot to go back and map those errors. Fixing now.

**Comment by carllerche** [comment]
So now it looks like azure is having problems reporting back status updates...

**Comment by carllerche** [comment]
Thanks :+1:

**Pull request merged**